File: php/MongoAnalyticsTest.php
<?php
require_once ('PHPUnit/Framework/Assert.php');
require_once ('PHPUnit/Framework/TestCase.php');

/**
 * Class MongoAnalyticsTest
 * Framework d'aggrégation et MapReduce pour manipuler des données analytiques
 */
class MongoAnalyticsTest extends PHPUnit_Framework_TestCase {

    private $mongoclient;
    private $db;
    private $hits;

    public function setUp(){
      $this->mongoclient = new MongoClient();
      $this->db = $this->mongoclient->selectDB("analytics");
      $this->hits = $this->db->selectCollection("hits");
    }

    public function tearDown(){
        $this->mongoclient->close();
    }

    /**
     * Trouver la plus grande affluence (nombre de requêtes par jour) pour l'url http://www.lateral-thoughts.com
     * avec le framework d'aggrégation
     */
    public function testFindHighestHitsForUrl(){
        // TODO définir le pipeline

        $result = $this->zips->aggregate();
        $this->assertEquals(66, $result['result'][0]['hits']);
        $this->assertEquals(2012, $result['result'][0]['_id']['y']);
        $this->assertEquals(3, $result['result'][0]['_id']['m']);
        $this->assertEquals(23, $result['result'][0]['_id']['d']);
    }

    /**
     * Trouver la plus grande affluence (nombre de requêtes par jour) pour l'url http://www.lateral-thoughts.com
     * avec un map reduce
     */
    public function testFindHighestHitsForUrlWithMapReduce(){
        $map = null;
        $reduce = null;

        $this->db->command(array(
            "mapreduce" => "hits",
            "map" => $map,
            "query" => array("url" => 'http://www.lateral-thoughts.com'),
            "reduce" => $reduce,
            "out" => "hits_per_url"));

        $hits_per_url = $this->db->selectCollection("hits_per_url");
        $query = array('_id.url'=>'http://www.lateral-thoughts.com');
        $this->assertTrue($hits_per_url->count($query)>0);
        $stats = $hits_per_url->find($query)->sort(array('value'=>-1))->limit(1);
        foreach($stats as $stat) {
            $this->assertEquals(66, $stat['value']);
            $this->assertEquals(2012, $stat['_id']['y']);
            $this->assertEquals(3, $stat['_id']['m']);
            $this->assertEquals(23, $stat['_id']['d']);
        }
    }
}
?>

File: python/MongoAnalyticsTest.py
# -*- coding: utf-8 -*-
import unittest
import pymongo
from pymongo import MongoClient
from bson.code import Code


class MongoAnalyticsTest(unittest.TestCase):

    #
    # Initialisation des membres privés à compléter
    def setUp(self):
        self.mongoclient = MongoClient()
        self.db = self.mongoclient.analytics
        self.hits = self.db.hits


    def tearDown(self): 
        self.mongoclient.close()


    #  Trouver la plus grande affluence (nombre de requêtes par jour) pour l'url http://www.lateral-thoughts.com
    # avec le framework d'aggrégation
    def testFindHighestHitsForUrl(self):
        match = {'$match' : {"url" : 'http://www.lateral-thoughts.com'}} 
        projection = {'$project' : 
                        {
                            '_id' : '$url', 
                            'date' : {
                                'y' : {'$year' : '$date'},
                                'm' : {'$month' : '$date'},
                                'd' : {'$dayOfMonth' : '$date'}
                            }
                        }
                      }
        group = {'$group' : {
                                '_id' : {
                                    'p' : '$_id',
                                    'y' : '$date.y',
                                    'm' : '$date.m',
                                    'd' : '$date.d'
                                }, 
                                'hits' : {'$sum' : 1}
                            }
                }
        
        sort = {"$sort" : {"hits" : -1}}
        limit = {"$limit" : 1}
        result = self.hits.aggregate([match, projection, group, sort, limit])
        self.assertEqual(result['result'][0]['hits'], 66)
        self.assertEqual(result['result'][0]['_id']['y'], 2012)
        self.assertEqual(result['result'][0]['_id']['m'], 3)
        self.assertEqual(result['result'][0]['_id']['d'], 23)

    #  Trouver la plus grande affluence (nombre de requêtes par jour) pour l'url http://www.lateral-thoughts.com
    # avec un map reduce
    def testFindHighestHitsForUrlWithMapReduce(self):
        mapper = Code("""
                       function() {
                        var key = {
                                    url : this.url,
                                    y : this.date.getFullYear(),
                                    m : this.date.getMonth()+1,
                                    d : this.date.getUTCDate(),
                                   };
                        emit(key,1);
                    }
                       """)
        reducer = Code("""
                       function(url, values) {
                            var res = 0;
                            return Array.sum(values)
                        }
                       """)


        result = self.db.hits.map_reduce(mapper, reducer, "hits_per_url")
        self.assertEqual(result.find({'_id.url':'http://www.lateral-thoughts.com'}).count(), 372)
        stats = result.find({'_id.url':'http://www.lateral-thoughts.com'}).sort([("value", -1)]).limit(1)
        for stat in stats:
            self.assertEqual(66, stat['value'])
            self.assertEqual(2012, stat['_id']['y'])
            self.assertEqual(3, stat['_id']['m'])
            self.assertEqual(23, stat['_id']['d'])

if __name__ == '__main__':
    unittest.main()
